// @antv/g-webgpu-core/lib/components/material/MaterialComponent.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MaterialComponent = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _ = require("../..");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var MaterialComponent = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(MaterialComponent, _Component);

  var _super = _createSuper(MaterialComponent);

  function MaterialComponent(data) {
    var _this;

    (0, _classCallCheck2.default)(this, MaterialComponent);
    _this = _super.call(this, data);
    _this.vertexShaderGLSL = void 0;
    _this.fragmentShaderGLSL = void 0;
    _this.dirty = true;
    _this.uniforms = [];
    _this.cull = {
      enable: true,
      face: _.gl.BACK
    };
    _this.depth = {
      enable: true
    };
    _this.blend = void 0;
    Object.assign((0, _assertThisInitialized2.default)(_this), data);
    return _this;
  }

  (0, _createClass2.default)(MaterialComponent, [{
    key: "setCull",
    value: function setCull(cull) {
      this.cull = cull;
      return this;
    }
  }, {
    key: "setDepth",
    value: function setDepth(depth) {
      this.depth = depth;
      return this;
    }
  }, {
    key: "setBlend",
    value: function setBlend(blend) {
      this.blend = blend;
      return this;
    }
  }, {
    key: "setUniform",
    value: function setUniform(name, data) {
      var _this2 = this;

      if (typeof name !== 'string') {
        Object.keys(name).forEach(function (key) {
          return _this2.setUniform(key, name[key]);
        });
        return this;
      }

      var existedUniform = this.uniforms.find(function (u) {
        return u.name === name;
      });

      if (!existedUniform) {
        this.uniforms.push({
          name: name,
          dirty: true,
          data: data
        });
      } else {
        existedUniform.dirty = true;
        existedUniform.data = data;
      }

      this.dirty = true;
      return this;
    }
  }]);
  return MaterialComponent;
}(_.Component);

exports.MaterialComponent = MaterialComponent;
