// @antv/g-webgpu-core/es/components/geometry/GeometryComponent.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import { Component } from '../..';
export var GeometryComponent = /*#__PURE__*/function (_Component) {
  _inherits(GeometryComponent, _Component);

  var _super = _createSuper(GeometryComponent);

  // instanced count
  function GeometryComponent(data) {
    var _this;

    _classCallCheck(this, GeometryComponent);

    _this = _super.call(this, data);
    _this.dirty = true;
    _this.attributes = [];
    _this.indices = void 0;
    _this.indicesBuffer = void 0;
    _this.vertexCount = void 0;
    _this.maxInstancedCount = void 0;
    _this.aabb = void 0;
    Object.assign(_assertThisInitialized(_this), data);
    return _this;
  }
  /**
   * @see https://threejs.org/docs/#api/en/core/BufferAttribute
   */


  _createClass(GeometryComponent, [{
    key: "setAttribute",
    value: function setAttribute(name, data, descriptor, bufferGetter) {
      var existed = this.attributes.find(function (a) {
        return a.name === name;
      });

      if (!existed) {
        this.attributes.push(_objectSpread(_objectSpread({
          dirty: true,
          name: name,
          data: data
        }, descriptor), {}, {
          bufferGetter: bufferGetter
        }));
      } else {
        existed.data = data;
        existed.dirty = true;
      }

      this.dirty = true;
      return this;
    }
  }, {
    key: "setIndex",
    value: function setIndex(data) {
      this.indices = new Uint32Array( // @ts-ignore
      data.buffer ? data.buffer : data);
      this.dirty = true;
      return this;
    }
  }]);

  return GeometryComponent;
}(Component);
