// @antv/g-webgpu-core/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Component", {
  enumerable: true,
  get: function get() {
    return _ComponentManager.Component;
  }
});
Object.defineProperty(exports, "ComponentManager", {
  enumerable: true,
  get: function get() {
    return _ComponentManager.ComponentManager;
  }
});
Object.defineProperty(exports, "FrameGraphSystem", {
  enumerable: true,
  get: function get() {
    return _System.FrameGraphSystem;
  }
});
Object.defineProperty(exports, "GeometryComponent", {
  enumerable: true,
  get: function get() {
    return _GeometryComponent.GeometryComponent;
  }
});
Object.defineProperty(exports, "IBoxGeometryParams", {
  enumerable: true,
  get: function get() {
    return _interface.IBoxGeometryParams;
  }
});
Object.defineProperty(exports, "GeometrySystem", {
  enumerable: true,
  get: function get() {
    return _System2.GeometrySystem;
  }
});
Object.defineProperty(exports, "IUniformBinding", {
  enumerable: true,
  get: function get() {
    return _interface2.IUniformBinding;
  }
});
Object.defineProperty(exports, "MaterialComponent", {
  enumerable: true,
  get: function get() {
    return _MaterialComponent.MaterialComponent;
  }
});
Object.defineProperty(exports, "MaterialSystem", {
  enumerable: true,
  get: function get() {
    return _System3.MaterialSystem;
  }
});
Object.defineProperty(exports, "CullableComponent", {
  enumerable: true,
  get: function get() {
    return _CullableComponent.CullableComponent;
  }
});
Object.defineProperty(exports, "IMeshParams", {
  enumerable: true,
  get: function get() {
    return _interface3.IMeshParams;
  }
});
Object.defineProperty(exports, "MeshComponent", {
  enumerable: true,
  get: function get() {
    return _MeshComponent.MeshComponent;
  }
});
Object.defineProperty(exports, "MeshSystem", {
  enumerable: true,
  get: function get() {
    return _System4.MeshSystem;
  }
});
Object.defineProperty(exports, "RendererSystem", {
  enumerable: true,
  get: function get() {
    return _System5.RendererSystem;
  }
});
Object.defineProperty(exports, "HierarchyComponent", {
  enumerable: true,
  get: function get() {
    return _HierarchyComponent.HierarchyComponent;
  }
});
Object.defineProperty(exports, "SceneGraphSystem", {
  enumerable: true,
  get: function get() {
    return _System6.SceneGraphSystem;
  }
});
Object.defineProperty(exports, "TransformComponent", {
  enumerable: true,
  get: function get() {
    return _TransformComponent.TransformComponent;
  }
});
Object.defineProperty(exports, "gl", {
  enumerable: true,
  get: function get() {
    return _gl.gl;
  }
});
Object.defineProperty(exports, "IAttribute", {
  enumerable: true,
  get: function get() {
    return _IAttribute.IAttribute;
  }
});
Object.defineProperty(exports, "IAttributeInitializationOptions", {
  enumerable: true,
  get: function get() {
    return _IAttribute.IAttributeInitializationOptions;
  }
});
Object.defineProperty(exports, "IBuffer", {
  enumerable: true,
  get: function get() {
    return _IBuffer.IBuffer;
  }
});
Object.defineProperty(exports, "IBufferInitializationOptions", {
  enumerable: true,
  get: function get() {
    return _IBuffer.IBufferInitializationOptions;
  }
});
Object.defineProperty(exports, "IComputeModel", {
  enumerable: true,
  get: function get() {
    return _IComputeModel.IComputeModel;
  }
});
Object.defineProperty(exports, "IElements", {
  enumerable: true,
  get: function get() {
    return _IElements.IElements;
  }
});
Object.defineProperty(exports, "IElementsInitializationOptions", {
  enumerable: true,
  get: function get() {
    return _IElements.IElementsInitializationOptions;
  }
});
Object.defineProperty(exports, "IFramebuffer", {
  enumerable: true,
  get: function get() {
    return _IFramebuffer.IFramebuffer;
  }
});
Object.defineProperty(exports, "IFramebufferInitializationOptions", {
  enumerable: true,
  get: function get() {
    return _IFramebuffer.IFramebufferInitializationOptions;
  }
});
Object.defineProperty(exports, "IModel", {
  enumerable: true,
  get: function get() {
    return _IModel.IModel;
  }
});
Object.defineProperty(exports, "IModelDrawOptions", {
  enumerable: true,
  get: function get() {
    return _IModel.IModelDrawOptions;
  }
});
Object.defineProperty(exports, "IModelInitializationOptions", {
  enumerable: true,
  get: function get() {
    return _IModel.IModelInitializationOptions;
  }
});
Object.defineProperty(exports, "IRenderbuffer", {
  enumerable: true,
  get: function get() {
    return _IRenderbuffer.IRenderbuffer;
  }
});
Object.defineProperty(exports, "IRenderbufferInitializationOptions", {
  enumerable: true,
  get: function get() {
    return _IRenderbuffer.IRenderbufferInitializationOptions;
  }
});
Object.defineProperty(exports, "BufferData", {
  enumerable: true,
  get: function get() {
    return _IRendererService.BufferData;
  }
});
Object.defineProperty(exports, "ICamera", {
  enumerable: true,
  get: function get() {
    return _IRendererService.ICamera;
  }
});
Object.defineProperty(exports, "IClearOptions", {
  enumerable: true,
  get: function get() {
    return _IRendererService.IClearOptions;
  }
});
Object.defineProperty(exports, "IReadPixelsOptions", {
  enumerable: true,
  get: function get() {
    return _IRendererService.IReadPixelsOptions;
  }
});
Object.defineProperty(exports, "IRendererConfig", {
  enumerable: true,
  get: function get() {
    return _IRendererService.IRendererConfig;
  }
});
Object.defineProperty(exports, "IRendererService", {
  enumerable: true,
  get: function get() {
    return _IRendererService.IRendererService;
  }
});
Object.defineProperty(exports, "IScene", {
  enumerable: true,
  get: function get() {
    return _IRendererService.IScene;
  }
});
Object.defineProperty(exports, "IView", {
  enumerable: true,
  get: function get() {
    return _IRendererService.IView;
  }
});
Object.defineProperty(exports, "IViewport", {
  enumerable: true,
  get: function get() {
    return _IRendererService.IViewport;
  }
});
Object.defineProperty(exports, "ITexture2D", {
  enumerable: true,
  get: function get() {
    return _ITexture2D.ITexture2D;
  }
});
Object.defineProperty(exports, "ITexture2DInitializationOptions", {
  enumerable: true,
  get: function get() {
    return _ITexture2D.ITexture2DInitializationOptions;
  }
});
Object.defineProperty(exports, "IUniform", {
  enumerable: true,
  get: function get() {
    return _IUniform.IUniform;
  }
});
Object.defineProperty(exports, "PixelPickingPass", {
  enumerable: true,
  get: function get() {
    return _PixelPickingPass.PixelPickingPass;
  }
});
Object.defineProperty(exports, "createEntity", {
  enumerable: true,
  get: function get() {
    return _Entity.createEntity;
  }
});
Object.defineProperty(exports, "IDENTIFIER", {
  enumerable: true,
  get: function get() {
    return _identifier.IDENTIFIER;
  }
});
Object.defineProperty(exports, "container", {
  enumerable: true,
  get: function get() {
    return _inversify.container;
  }
});
Object.defineProperty(exports, "createWorldContainer", {
  enumerable: true,
  get: function get() {
    return _inversify.createWorldContainer;
  }
});
Object.defineProperty(exports, "lazyInject", {
  enumerable: true,
  get: function get() {
    return _inversify.lazyInject;
  }
});
Object.defineProperty(exports, "lazyMultiInject", {
  enumerable: true,
  get: function get() {
    return _inversify.lazyMultiInject;
  }
});
Object.defineProperty(exports, "ISystem", {
  enumerable: true,
  get: function get() {
    return _ISystem.ISystem;
  }
});
Object.defineProperty(exports, "IConfig", {
  enumerable: true,
  get: function get() {
    return _IConfigService.IConfig;
  }
});
Object.defineProperty(exports, "IConfigService", {
  enumerable: true,
  get: function get() {
    return _IConfigService.IConfigService;
  }
});
Object.defineProperty(exports, "IInteractorEvent", {
  enumerable: true,
  get: function get() {
    return _IIteractorService.IInteractorEvent;
  }
});
Object.defineProperty(exports, "IInteractorService", {
  enumerable: true,
  get: function get() {
    return _IIteractorService.IInteractorService;
  }
});
Object.defineProperty(exports, "IShaderModuleService", {
  enumerable: true,
  get: function get() {
    return _IShaderModuleService.IShaderModuleService;
  }
});
Object.defineProperty(exports, "Frustum", {
  enumerable: true,
  get: function get() {
    return _Frustum.Frustum;
  }
});
Object.defineProperty(exports, "isSafari", {
  enumerable: true,
  get: function get() {
    return _isSafari.isSafari;
  }
});
exports.DefineValuePlaceholder = exports.Target = exports.STORAGE_CLASS = exports.AST_NODE_TYPES = exports.AST_TOKEN_TYPES = void 0;

require("reflect-metadata");

var _ComponentManager = require("./ComponentManager");

var _System = require("./components/framegraph/System");

var _GeometryComponent = require("./components/geometry/GeometryComponent");

var _interface = require("./components/geometry/interface");

var _System2 = require("./components/geometry/System");

var _interface2 = require("./components/material/interface");

var _MaterialComponent = require("./components/material/MaterialComponent");

var _System3 = require("./components/material/System");

var _CullableComponent = require("./components/mesh/CullableComponent");

var _interface3 = require("./components/mesh/interface");

var _MeshComponent = require("./components/mesh/MeshComponent");

var _System4 = require("./components/mesh/System");

var _System5 = require("./components/renderer/System");

var _HierarchyComponent = require("./components/scenegraph/HierarchyComponent");

var _System6 = require("./components/scenegraph/System");

var _TransformComponent = require("./components/scenegraph/TransformComponent");

var _gl = require("./components/renderer/gl");

var _IAttribute = require("./components/renderer/IAttribute");

var _IBuffer = require("./components/renderer/IBuffer");

var _IComputeModel = require("./components/renderer/IComputeModel");

var _IElements = require("./components/renderer/IElements");

var _IFramebuffer = require("./components/renderer/IFramebuffer");

var _IModel = require("./components/renderer/IModel");

var _IRenderbuffer = require("./components/renderer/IRenderbuffer");

var _IRendererService = require("./components/renderer/IRendererService");

var _ITexture2D = require("./components/renderer/ITexture2D");

var _IUniform = require("./components/renderer/IUniform");

var _PixelPickingPass = require("./components/renderer/passes/PixelPickingPass");

var _Entity = require("./Entity");

var _identifier = require("./identifier");

var _inversify = require("./inversify.config");

var _ISystem = require("./ISystem");

var _IConfigService = require("./services/config/IConfigService");

var _IIteractorService = require("./services/interactor/IIteractorService");

var _IShaderModuleService = require("./services/shader-module/IShaderModuleService");

var _Frustum = require("./shape/Frustum");

var _isSafari = require("./utils/isSafari");

var AST_TOKEN_TYPES;
exports.AST_TOKEN_TYPES = AST_TOKEN_TYPES;

(function (AST_TOKEN_TYPES) {
  AST_TOKEN_TYPES["Void"] = "Void";
  AST_TOKEN_TYPES["Boolean"] = "Boolean";
  AST_TOKEN_TYPES["Float"] = "Float";
  AST_TOKEN_TYPES["Uint32"] = "Uint32";
  AST_TOKEN_TYPES["Int32"] = "Int32";
  AST_TOKEN_TYPES["Vector"] = "Vector";
  AST_TOKEN_TYPES["Vector2Float"] = "vec2<f32>";
  AST_TOKEN_TYPES["Vector3Float"] = "vec3<f32>";
  AST_TOKEN_TYPES["Vector4Float"] = "vec4<f32>";
  AST_TOKEN_TYPES["Vector2Boolean"] = "vec2<bool>";
  AST_TOKEN_TYPES["Vector3Boolean"] = "vec3<bool>";
  AST_TOKEN_TYPES["Vector4Boolean"] = "vec4<bool>";
  AST_TOKEN_TYPES["Vector2Uint"] = "vec2<u32>";
  AST_TOKEN_TYPES["Vector3Uint"] = "vec3<u32>";
  AST_TOKEN_TYPES["Vector4Uint"] = "vec4<u32>";
  AST_TOKEN_TYPES["Vector2Int"] = "vec2<i32>";
  AST_TOKEN_TYPES["Vector3Int"] = "vec3<i32>";
  AST_TOKEN_TYPES["Vector4Int"] = "vec4<i32>";
  AST_TOKEN_TYPES["Matrix"] = "Matrix";
  AST_TOKEN_TYPES["Matrix3x3Float"] = "mat3x3<f32>";
  AST_TOKEN_TYPES["Matrix4x4Float"] = "mat4x4<i32>";
  AST_TOKEN_TYPES["Struct"] = "Struct";
  AST_TOKEN_TYPES["FloatArray"] = "Float[]";
  AST_TOKEN_TYPES["Vector4FloatArray"] = "vec4<f32>[]";
})(AST_TOKEN_TYPES || (exports.AST_TOKEN_TYPES = AST_TOKEN_TYPES = {}));

var AST_NODE_TYPES;
exports.AST_NODE_TYPES = AST_NODE_TYPES;

(function (AST_NODE_TYPES) {
  AST_NODE_TYPES["Program"] = "Program";
  AST_NODE_TYPES["Identifier"] = "Identifier";
  AST_NODE_TYPES["VariableDeclaration"] = "VariableDeclaration";
  AST_NODE_TYPES["BlockStatement"] = "BlockStatement";
  AST_NODE_TYPES["ReturnStatement"] = "ReturnStatement";
  AST_NODE_TYPES["FunctionDeclaration"] = "FunctionDeclaration";
  AST_NODE_TYPES["VariableDeclarator"] = "VariableDeclarator";
  AST_NODE_TYPES["AssignmentExpression"] = "AssignmentExpression";
  AST_NODE_TYPES["LogicalExpression"] = "LogicalExpression";
  AST_NODE_TYPES["BinaryExpression"] = "BinaryExpression";
  AST_NODE_TYPES["ArrayExpression"] = "ArrayExpression";
  AST_NODE_TYPES["UnaryExpression"] = "UnaryExpression";
  AST_NODE_TYPES["UpdateExpression"] = "UpdateExpression";
  AST_NODE_TYPES["FunctionExpression"] = "FunctionExpression";
  AST_NODE_TYPES["MemberExpression"] = "MemberExpression";
  AST_NODE_TYPES["ConditionalExpression"] = "ConditionalExpression";
  AST_NODE_TYPES["ExpressionStatement"] = "ExpressionStatement";
  AST_NODE_TYPES["CallExpression"] = "CallExpression";
  AST_NODE_TYPES["NumThreadStatement"] = "NumThreadStatement";
  AST_NODE_TYPES["StorageStatement"] = "StorageStatement";
  AST_NODE_TYPES["DoWhileStatement"] = "DoWhileStatement";
  AST_NODE_TYPES["WhileStatement"] = "WhileStatement";
  AST_NODE_TYPES["ForStatement"] = "ForStatement";
  AST_NODE_TYPES["BreakStatement"] = "BreakStatement";
  AST_NODE_TYPES["ContinueStatement"] = "ContinueStatement";
  AST_NODE_TYPES["IfStatement"] = "IfStatement";
  AST_NODE_TYPES["ImportedFunctionStatement"] = "ImportedFunctionStatement";
})(AST_NODE_TYPES || (exports.AST_NODE_TYPES = AST_NODE_TYPES = {}));

var STORAGE_CLASS;
exports.STORAGE_CLASS = STORAGE_CLASS;

(function (STORAGE_CLASS) {
  STORAGE_CLASS["Input"] = "Input";
  STORAGE_CLASS["Output"] = "Output";
  STORAGE_CLASS["Uniform"] = "Uniform";
  STORAGE_CLASS["Workgroup"] = "Workgroup";
  STORAGE_CLASS["UniformConstant"] = "UniformConstant";
  STORAGE_CLASS["Image"] = "Image";
  STORAGE_CLASS["StorageBuffer"] = "StorageBuffer";
  STORAGE_CLASS["Private"] = "Private";
  STORAGE_CLASS["Function"] = "Function";
})(STORAGE_CLASS || (exports.STORAGE_CLASS = STORAGE_CLASS = {}));

/**
 * 根据目标平台生成 Shader 代码
 * * WebGL GLSL 1.0
 * * WebGPU Chrome/Edge GLSL 4.5 & WGSL @see https://gpuweb.github.io/gpuweb/wgsl.html
 * * Safari WHLSL (maybe deprecated)
 */
var Target;
exports.Target = Target;

(function (Target) {
  Target["GLSL100"] = "GLSL100";
  Target["GLSL450"] = "GLSL450";
  Target["WGSL"] = "WGSL";
})(Target || (exports.Target = Target = {}));

var DefineValuePlaceholder = '__DefineValuePlaceholder__';
exports.DefineValuePlaceholder = DefineValuePlaceholder;
